x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceStats.java

```diff
@@ -226,22 +234,52 @@ public Accumulator merge(InferenceStats otherStats) {
         }
 
         public Accumulator incMissingFields() {
-            this.missingFieldsAccumulator.increment();
-            return this;
+            readWriteLock.readLock().lock();
+            try {
+                this.missingFieldsAccumulator.increment();
+                return this;
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
         }
 
         public Accumulator incInference() {
-            this.inferenceAccumulator.increment();
-            return this;
+            readWriteLock.readLock().lock();
+            try {
+                this.inferenceAccumulator.increment();
+                return this;
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
         }
 
         public Accumulator incFailure() {
-            this.failureCountAccumulator.increment();
-            return this;
+            readWriteLock.readLock().lock();
+            try {
+                this.failureCountAccumulator.increment();
+                return this;
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
         }
 
-        public InferenceStats currentStats() {
-            return currentStats(Instant.now());
+        /**
+         * Thread safe.
+         *
+         * Returns the current stats and resets the values of all the counters.
+         * @return The current stats
+         */
+        public InferenceStats currentStatsAndReset() {
+            readWriteLock.writeLock().lock();
+            try {
+                InferenceStats stats = currentStats(Instant.now());
```

Comment:
nit: make `public InferenceStats currentStats(Instant timeStamp)` private as it is only called here and not protected by the lock otherwise

Reply:
It is called in the stats persister when merging stats. 

I will add a comment about it not being thread safe.

Reply:
Actually, it is thread safe and it is used in the stats service.

getting the `longValue` without resetting is perfectly safe. Its guaranteed to get the latest count on the stats. 

And since where it is used, no more updates are occurring, it should be safe.